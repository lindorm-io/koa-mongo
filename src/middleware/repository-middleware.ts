import { IKoaMongoContext } from "../types";
import { Middleware, Next } from "koa";
import { RepositoryBase } from "@lindorm-io/mongo";
import { camelCase } from "lodash";

interface RepositoryMiddlewareOptions {
  key?: string;
}

export const repositoryMiddleware = (
  Repository: typeof RepositoryBase,
  options?: RepositoryMiddlewareOptions,
): Middleware => async (ctx: IKoaMongoContext, next: Next): Promise<void> => {
  const start = Date.now();

  const db = await ctx.mongo.getDatabase();
  const logger = ctx.logger;

  /*
   * Ignoring TS here since Repository needs to be abstract
   * to ensure that all input at least attempts to be unique
   */
  // @ts-ignore
  ctx.repository[camelCase(options?.key || Repository.name)] = new Repository({ db, logger });

  ctx.metrics.repository = (ctx.metrics.repository || 0) + (Date.now() - start);

  await next();
};
