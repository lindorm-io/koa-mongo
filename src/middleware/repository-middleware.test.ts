import { repositoryMiddleware } from "./repository-middleware";
import { logger } from "../test";

class Test {}

describe("repositoryMiddleware", () => {
  let ctx: any;
  let next: any;

  beforeEach(() => {
    ctx = {
      logger,
      metrics: {},
      mongo: { getDatabase: () => "db" },
      repository: {},
    };
    next = () => Promise.resolve();
  });

  test("should set repository on context", async () => {
    // @ts-ignore
    await expect(repositoryMiddleware(Test)(ctx, next)).resolves.toBe(undefined);
    expect(ctx.repository).toMatchSnapshot();
    expect(ctx.metrics.repository).toStrictEqual(expect.any(Number));
  });

  test("should set repository with specific key", async () => {
    // @ts-ignore
    await expect(repositoryMiddleware(Test, { key: "otherKey" })(ctx, next)).resolves.toBe(undefined);
    expect(ctx.repository).toMatchSnapshot();
    expect(ctx.metrics.repository).toStrictEqual(expect.any(Number));
  });
});
