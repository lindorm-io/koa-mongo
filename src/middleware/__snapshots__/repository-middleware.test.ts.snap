// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`repositoryMiddleware should set repository on context 1`] = `
Object {
  "test": Test {},
}
`;

exports[`repositoryMiddleware should set repository with specific key 1`] = `
Object {
  "otherKey": Test {},
}
`;
